File: Java/math/tribonacci.java
/*    Tribonacci Series:
   Tribonacci series is a generalisation of Fibonacci series where current element(greater than 0)
   is the sum of preceding three elements,
   i.e. T(n)=T(n-1)+T(n-2)+T(n-3).

      Approach:

   The sum of three preceding variables will be stored in "sum" variable.
   Variable "first" will be overrided by "second" which in turn will be overrided by "third" which
   will be overrided by "sum".
   This process will continue till n terms.  */

import java.util.Scanner;

public class Main {
    public static void main(String[] args) {
        Scanner scan = new Scanner(System.in);
        long num;
        System.out.println("Enter the number upto which the series is to be printed");
        num = scan.nextLong();
        printTribonacci(num);
    }

    public static void printTribonacci(long num) {
        long first = 0, second = 0, third = 1, sum = 0, count = 0;
        while (num > 0) {
            count++;
            if (count == 1)
                System.out.print(0);
            else if (count == 2)
                System.out.print(", " + 0);
            else if (count == 3)
                System.out.print(", " + 1);
            else {
                sum = first + second + third;
                System.out.print(", " + sum);
                //Storing the values of succeeding terms in respective variables.
                first = second;
                second = third;
                third = sum;
            }
            num--;    //Decreasing the value of total number of terms by one after each iteration.
        }
    }
}

 /*                                              OUTPUT:
 Test Case 1:
 Enter the number upto which the series is to be printed
 23
 0, 0, 1, 1, 2, 4, 7, 13, 24, 44, 81, 149, 274, 504, 927, 1705, 3136, 5768, 10609, 19513, 35890, 66012, 121415

 Space Complexity: O(1)
 Time Complexity: O(n)
 */

File: Julia/math/tribonacci.jl
"""Julia program to find the n'th number in the tribonacci series
Tribonacci series is a generalization of the Fibonacci sequence, in which the current term
is the sum of the previous three 
"""

function find_tribonacci(n)
    dp = zeros(Int, n + 1)
    dp[1] = 0
    dp[2] = 0
    dp[3] = 1

    for i in 4:n
        dp[i] = dp[i-1] + dp[i-2] + dp[i-3]
    end

    return dp[n]
end


print("Enter the value of n?, where you need the n'th number in the tribonacci sequence. ")
n = readline()
n = parse(Int, n)
if (n <= 0)
    println("The given value of n is invalid.")
    exit()
end
res = find_tribonacci(n)
println("The $n'th term in the tribonacci series is $res.")


"""
Time Complexity - O(n)
Space Complexity - O(n)

SAMPLE INPUT AND OUTPUT

SAMPLE I
Enter the value of n?, where you need the n'th number in the tribonacci sequence. 12
The 12'th term in the tribonacci series is 149.

SAMPLE II
Enter the value of n?, where you need the n'th number in the tribonacci sequence. 1254
The 1254'th term in the tribonacci series is 4020147461713125140.
"""
